### PR TITLE
DTSPO-75: added validation checks and variable

### PR DESCRIPTION
**Pull request by mokainos**
https://tools.hmcts.net/jira/browse/DTSPO-75

### Change description ###

1) Added Validation step and plan
2)  Added kv resource group variable  

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x ] No
```

### CODE REVIEW

**Review by dawsonrs — APPROVED** [review]

File: frontdoor.tf [hunk 1]
```diff
@@ -1,6 +1,6 @@
 resource "azurerm_frontdoor" "main" {
-  name                                         = "${var.project}-${var.env}"
-  location                                     = "global"
+  name = "${var.project}-${var.env}"
+  #location                                     = "global"
```
Comment (timja) [inline frontdoor.tf:3]: hmm will this break existing deployed ones? probably not but you shouldn't leave commented out code

```suggestion
```

File: main.tf [hunk 2]
```diff
@@ -0,0 +1,7 @@
+/**
```
Comment (timja) [inline main.tf:1]: is this file required?

File: validation/fd.tf [hunk 3]
```diff
@@ -0,0 +1,24 @@
+module "landing_zone" {
+  source = "git::https://github.com/hmcts/terraform-module-frontdoor.git?ref=features/dtspo75_1"
```
Comment (timja) [inline validation/fd.tf:2]: make sure to update after merge

File: validation/provider.tf [hunk 4]
```diff
@@ -0,0 +1,14 @@
+provider "azurerm" {
+  alias = "data"
+  features {}
+  subscription_id = "bf308a5c-0624-4334-8ff8-8dca9fd43783"
+}
+
+
+provider "azurerm" {
+  #skip_provider_registration = "true"
```
Comment (timja) [inline validation/provider.tf:9]:
```suggestion
```

File: validation/provider.tf [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+  #version         = "= 2.4.0"
```
Comment (timja) [inline validation/provider.tf:10]: either choose a version or lock it
```suggestion
```

both have ±'s

File: variables.tf [hunk 6]
```diff
@@ -1,13 +1,58 @@
...
-variable "resource_group" {}
-variable "certificate_key_vault_name" {}
-variable "oms_env" {}
-variable "certificate_name_check" {}
+variable "env" {
+  description = "Enter name of the enviornment to deploy frontdoor"
+  type        = string
+}
+variable "subscription" {
+  description = "Enter name of the subsription to deploy frontdoor"
```
Comment (timja) [inline variables.tf:6]:
```suggestion
  description = "Name of the subscription to deploy frontdoor, e.g. stg"
```

File: variables.tf [hunk 7]
```diff
@@ -1,13 +1,58 @@
...
-variable "frontends" {}
-variable "subscription_id" {}
-variable "enable_ssl" {}
-variable "ssl_mode" {}
-variable "resource_group" {}
-variable "certificate_key_vault_name" {}
-variable "oms_env" {}
-variable "certificate_name_check" {}
+variable "env" {
+  description = "Enter name of the enviornment to deploy frontdoor"
```
Comment (timja) [inline variables.tf:2]:
```suggestion
  description = "Enter name of the environment to deploy frontdoor"
```

File: variables.tf [hunk 8]
```diff
@@ -1,13 +1,58 @@
...
+  type        = string
+}
+variable "subscription" {
+  description = "Enter name of the subsription to deploy frontdoor"
+  type        = string
+}
+variable "project" {
+  description = "Enter Name of the project"
```
Comment (timja) [inline variables.tf:10]:
```suggestion
  description = "Name of the project"
```

File: variables.tf [hunk 9]
```diff
@@ -1,13 +1,58 @@
...
+  type        = string
+  default     = "hmcts"
+}
+variable "location" {
+  description = "Enter Azure location to deploy the resource"
```
Comment (timja) [inline variables.tf:15]:
```suggestion
  description = "Azure location to deploy the resource"
```

File: variables.tf [hunk 10]
```diff
@@ -1,13 +1,58 @@
...
+variable "subscription" {
+  description = "Enter name of the subsription to deploy frontdoor"
+  type        = string
+}
+variable "project" {
+  description = "Enter Name of the project"
+  type        = string
+  default     = "hmcts"
+}
+variable "location" {
```
Comment (timja) [inline variables.tf:14]: We need to see if we actually need this now

File: variables.tf [hunk 11]
```diff
@@ -1,13 +1,58 @@
...
+variable "subscription_id" {
+  description = "Enter Subscription ID"
+  type        = string
+}
+variable "enable_ssl" {
+  description = "Enable SSL"
+  type        = bool
+}
+variable "ssl_mode" {
+  description = "Certificate source to encrypted HTTPS traffic with"
```
Comment (timja) [inline variables.tf:36]: please specify the allowed values, AzureKeyVault is one I believe and Managed might be the other but look it up
```suggestion
  description = "Certificate source to encrypt HTTPS traffic with"
```

File: variables.tf [hunk 12]
```diff
@@ -1,13 +1,58 @@
...
+variable "resource_group" {
+  description = "Enter Resource Group Name"
+  type        = string
+}
+variable "certificate_key_vault_name" {
+  description = "Name of the Keyvault that holds certificate"
+  type        = string
+}
+variable "oms_env" {
+  description = "Name of the Monitoring enviornment"
```
Comment (timja) [inline variables.tf:48]:
```suggestion
  description = "Name of the log analytics workspace"
```

File: variables.tf [hunk 13]
```diff
@@ -1,13 +1,58 @@
...
+  type        = string
+}
+variable "certificate_name_check" {
+  description = "Enforce_backend_pools_certificate_name_check"
```
Comment (timja) [inline variables.tf:52]:
```suggestion
  description = "Enforce backend pools certificate name check"
```

File: variables.tf [hunk 14]
```diff
@@ -1,13 +1,58 @@
...
+  type        = bool
+}
+variable "kv_resource_group" {
```
Comment (timja) [inline variables.tf:55]: we haven't really abbreviated any other variables, this reads better
```suggestion
variable "key_vault_resource_group" {
```

File: variables.tf [hunk 15]
```diff
@@ -1,13 +1,58 @@
...
+  description = "Keyvault resource group name"
```
Comment (timja) [inline variables.tf:56]: key vault has spaces in it's name https://azure.microsoft.com/en-gb/services/key-vault/
```suggestion
  description = "Key Vault resource group name"
```

**Review by timja — CHANGES_REQUESTED** [review]
remove the commented out code and fix the typoes please,

would be good if we can verify what happens if we remove the location variable

File: frontdoor.tf (hunk 1 above)
Comment (mokainos) [reply frontdoor.tf:3]: Location argument is deprecated and defaults to "global" now

Comment (timja) [reply frontdoor.tf:3]: so remove then assuming the existing frontdoors work

Comment (mokainos) [reply frontdoor.tf]: removed it

**Comment by naikajah** [comment]
Guys!! there are hell lot of issues with Terraform for Azure Frontdoor and it seems, its not going to get fixed anytime soon because of the way frontdoor APIs are designed. I suggest we also look into ARM templates for frontdoor specifically for Frontdoor. As ARM templates are much better to work with in case of frontdoor.

File: main.tf (hunk 2 above)
Comment (mokainos) [reply main.tf:1]: yes to add text to  README.md using terraform-docs

File: validation/fd.tf (hunk 3 above)
Comment (mokainos) [reply validation/fd.tf:2]: will do

File: validation/provider.tf (hunk 5 above)
Comment (mokainos) [reply validation/provider.tf]: Deleted them.

File: validation/provider.tf (hunk 4 above)
Comment (mokainos) [reply validation/provider.tf]: Deleted them.

File: README.md [hunk 16]
```diff
@@ -1 +1,38 @@
...
+
+| Name | Description | Type | Default | Required |
+|------|-------------|------|---------|:--------:|
+| certificate\_key\_vault\_name | Name of the Keyvault that holds certificate | `string` | n/a | yes |
+| certificate\_name\_check | Enforce\_backend\_pools\_certificate\_name\_check | `bool` | n/a | yes |
+| common\_tags | Common tag to be applied | `map(string)` | n/a | yes |
+| enable\_ssl | Enable SSL | `bool` | n/a | yes |
+| env | Enter name of the enviornment to deploy frontdoor | `string` | n/a | yes |
+| frontends | Variable holds frontdoor configuration | `any` | n/a | yes |
+| kv\_resource\_group | Keyvault resource group name | `string` | n/a | yes |
```
Comment (timja) [inline README.md:26]: looks like the readme needs regenerating, this is the old var name

File: variables.tf (hunk 10 above)
Comment (mokainos) [reply variables.tf:14]: Removed default entry

File: README.md (hunk 16 above)
Comment (mokainos) [reply README.md]: updated README.md

**Review by timja — APPROVED** [review]

**Pull request merged**